lib/forms.js

```diff
@@ -11,55 +11,136 @@ var deleteAppReferences = require('./impl/deleteAppRefrences.js');
+  }
+
+  //There is already a mongoose connection, can already use this.
+  if (connections.mongooseConnection) {
+    logger.debug("Using cached mongoose Db connection");
+    return cb(undefined, connections.mongooseConnection);
+  }
+
+  var dbName = parsedMongoUri.database;
+
+  //Already have a mongoose connection, need to get a handle to the other database.
+  var mongooseDbConnection = mongooseConnection.useDb(dbName);
+
+  logger.debug("Mongoose connected to new db");
+  //Authenticating to the database
+
+  connections.mongooseConnection = mongooseDbConnection;
+
+  //Db is authenticated, can now do stuff.
+  return cb(undefined, mongooseDbConnection);
+}
 
 
 var forms = {
   connections: {},
   setConfig: config.set,
 
-  initConnection : function(options, cb) {
+  initConnection : function(options, callback) {
     var self = this;
     var uri = options.uri;
     var key = options.key || uri;
     var config = options.config || {};
-    if ( self.connections[key] ) {
-      return cb(undefined, self.connections[key]);
-    }
-    var paramValidation = validate(options);
 
-    logger.info("Initialising mongo connections");
 
-    function handleConnectionResponse(err, databaseConn) {
-      if (err) {
-        logger.error("Error connecting to mongo using client", err);
-        return cb(err);
-      }
+    //If there is already a mongoose connection, we can use the shared pool to connect to it.
+    //Prevents having too many mongo connections open.
+
+    async.waterfall([
+      function validateParams(cb) {
+        logger.debug("Validate Params Start", options);
+        var paramValidation = validate(options);
 
-      logger.info("Connected to mongo");
+        paramValidation.has("uri", function(failed) {
+          if (failed) {
+            return cb(new Error("Invalid Params: " + JSON.stringify(failed)));
+          }
 
-      mongoose.Promise = global.Promise;
-      var connections =  {
-        "databaseConnection": databaseConn,
-        "mongooseConnection": mongoose.createConnection(uri)
-      };
+          logger.debug("Valid Params");
 
-      self.connections[key] = connections;
+          return cb();
+        });
+      },
+      //First getting a mongoose db connection.
+      function getMongooseConnectionDb(cb) {
+
+        logger.debug("Getting Mongoose Connection");
+        self.connections[key] = self.connections[key] || {};
+        getMongooseConnection(uri, self.connections[key], cb);
+      },
+      //Getting a mongo db connection (for gridfs file operations)
+      function getDbConnection(mongooseConnection, cb) {
+        logger.debug("Got Mongoose Connection. Getting DB Connection for GridFS");
+
+
+        //If there is already a database connection for this key, don't try and connect again.
+        if (self.connections[key] && self.connections[key].databaseConnection) {
+          return cb(undefined, self.connections[key]);
+        }
 
+        //TODO: Clean this.
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)